pkg/deploy/controller.go

```diff
@@ -15,37 +15,48 @@ import (
 	"time"
 
 	errors2 "github.com/pkg/errors"
-	v12 "github.com/rancher/k3s/pkg/apis/k3s.cattle.io/v1"
-	v1 "github.com/rancher/k3s/pkg/generated/controllers/k3s.cattle.io/v1"
+	apisv1 "github.com/rancher/k3s/pkg/apis/k3s.cattle.io/v1"
+	controllersv1 "github.com/rancher/k3s/pkg/generated/controllers/k3s.cattle.io/v1"
 	"github.com/rancher/wrangler/pkg/apply"
 	"github.com/rancher/wrangler/pkg/merr"
 	"github.com/rancher/wrangler/pkg/objectset"
+	"github.com/rancher/wrangler/pkg/schemes"
 	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	yamlDecoder "k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/kubernetes"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/record"
 )
 
 const (
-	ns       = "kube-system"
-	startKey = "_start_"
+	startKey      = "_start_"
+	componentName = "deploy-controller"
 )
 
-func WatchFiles(ctx context.Context, apply apply.Apply, addons v1.AddonController, disables map[string]bool, bases ...string) error {
+func WatchFiles(ctx context.Context, client kubernetes.Interface, apply apply.Apply, addons controllersv1.AddonController, disables map[string]bool, bases ...string) error {
+	nodeName := os.Getenv("NODE_NAME")
```

Comment:
This code-path is spun up in `server.Run()` prior to the invocation of `agent.Run()`, which sets the `NODE_NAME` envvar, so we have a race to use a value that may not yet be available. 

Consider emulating what the etcd controller is doing in the sync loop: https://github.com/rancher/k3s/blob/master/pkg/etcd/controller.go#L43-L48 (but beware that this may introduce deadlocking).

Reply:
Good catch. I deferred the recorder setup until later, right before we list files for the first time.